kinds/report.go

```diff
@@ -0,0 +1,113 @@
+//go:build ignore
+// +build ignore
+
+//go:generate go run report.go
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+
+	"github.com/grafana/codejen"
+	"github.com/grafana/grafana/pkg/kindsys"
+	"github.com/grafana/grafana/pkg/registry/corekind"
+)
+
+func main() {
+	kinds := loadKinds()
+
+	sort.Slice(kinds, func(i, j int) bool {
+		return kinds[i].Name < kinds[j].Name
+	})
+
+	out := elsedie(json.MarshalIndent(kinds, "", "    "))("error generating json output")
+
+	path := filepath.Join(kindsys.DeclParentPath, "report.json")
+	file := codejen.NewFile(path, out, reportJenny{})
+	filesystem := elsedie(file.ToFS())("error building in-memory file system")
+
+	cwd := elsedie(os.Getwd())("error getting working directory")
+	groot := filepath.Dir(cwd)
+
+	if _, set := os.LookupEnv("CODEGEN_VERIFY"); set {
+		if err := filesystem.Verify(context.Background(), groot); err != nil {
+			die(fmt.Errorf("generated code is out of sync with inputs:\n%s\nrun `make gen-cue` to regenerate", err))
+		}
+	} else if err := filesystem.Write(context.Background(), groot); err != nil {
+		die(fmt.Errorf("error while writing generated code to disk:\n%s", err))
+	}
+}
+
+type kind struct {
+	Name     string `json:"name"`
+	Maturity string `json:"maturity"`
+	Type     string `json:"type"`
+	Latest   string `json:"latest,omitempty"`
+}
+
+func loadKinds() []kind {
+	return append(loadStructuredKinds(), loadRawKinds()...)
+}
+
+func loadStructuredKinds() []kind {
+	b := corekind.NewBase(nil)
+	allStructured := b.AllStructured()
+	kinds := make([]kind, 0, len(allStructured))
+
+	for _, k := range allStructured {
+		kinds = append(kinds, kind{
+			Name:     k.MachineName(),
+			Maturity: k.Maturity().String(),
+			Type:     "structured",
```

Comment:
Yeah it's not directly exposed - i opted to just have it be implicitly part of the kind category.

Reply:
IMO this would be adequately handled by opting to have the root output be an object with keys:

```json
{
  "core": [ ... ],
  "raw": [ ... ],
  "composable": [ ... ],
}
```

That's basically how we're doing it elsewhere. That also means that the type for each elem is consistent, and gives us space to add the plugin-generated kind info later.

Reply:
I guess it makes sense, yeah! 

Then, I suspect something like what I suggested at https://github.com/grafana/grafana/pull/59742#discussion_r1039933782 is the way to go because we'd still need to have a way to differentiate between kind types, right? 🤔

Reply:
Yep! Splitting them up that way should mean you could easily populate a struct of this type:

```go
type KindStateReport struct {
	Core       []kindsys.CoreStructuredProperties `json:"core"`
	Raw        []kindsys.RawProperties            `json:"raw"`
	Composable []kindsys.ComposableProperties     `json:"composable"`
}
```